Python/ceval.c

```diff
@@ -2121,9 +2121,26 @@ _PyEval_UnpackIterableStackRef(PyThreadState *tstate, _PyStackRef v_stackref,
             return 1;
         }
         Py_DECREF(w);
-        _PyErr_Format(tstate, PyExc_ValueError,
-                      "too many values to unpack (expected %d)",
-                      argcnt);
+
+        if (PyList_CheckExact(v) || PyTuple_CheckExact(v)
+              || PyDict_CheckExact(v)) {
+            ll = PyDict_CheckExact(v) ? PyDict_Size(v) : Py_SIZE(v);
```

Comment:
@encukou I think it makes sense to broaden the support to all sequences and mappings then:
```suggestion
        if (PySequence_Check(v) || PyMapping_Check(v)) {
            ll = PyObject_Size(v);
```

Reply:
The reason why we don't want to is to avoid calling arbitrary `__len__`. Otherwise, my suggestion would have just to use `_PyObject_HasLen` to determine whether `__len__` exists or not (see https://github.com/python/cpython/pull/122244#discussion_r1728522434).

Reply:
Yeah, we went there and back again.

Reply:
I think we're in good shape for now.

Reply:
Ah I see. Well I personally think it's really the user's responsibility to not make an evil `__len__`, but I understand the decision.